Fix duplicates in paging due to cost truncation in token

t2gran opened the pull request:
### Summary

- Fixed rounding errors in `Cost.toSeconds()` and `TimeAndCost.toSeconds()` methods. These should round to 
   closest cost-second, not truncate. This changed the cost by 1 in some of the SnapShot tests - this is expected.
- Updated itinerary and leg models to use the corrected rounding methods for time and cost. This needs to be consistent with the values saved in the paging token, and the values used during itinerary filtering for sorting itineraries. Times are truncated, and costs are rounded to nearest second. 

### Issue

The issue I was able to reproduce was cased by truncating the cost when creating the paging-token, this "shifted" the actual cost of the itinerary used to crop the next search slightly. So, when the exact same itinerary is found, then the itinerary was kept because it is sorted after the value in the page-token. To fix the problem I made sure all times and costs are normalized BEFORE we use them in the filter-chain and before the paging token is created. This will ensure that the fix works for both sort-directions (arrive-by and depart-after). I could not recreate this issue with the start/end time of the itineraries - but I think that is pure luck and not because the code was designed for it. So, I normalized both the times and cost to prevent this from happening in the future.

Note! This can not be fixed by rounding up/down because we do a `<=` or `>=` compare depending on the search direction. We could account for this inside the sort, but that would be super messy. Another solution would be to change the paging - but is difficult require some careful replacement of the existing values in it. 

Closes #6926

### Unit tests

I did write a integration test for this to reproduce the problem - but it is so complicated/ugly that I deleted it. Instead I have added a few unit-tests, and a comment in the code to remind us to be careful when changing the paging token. This solution is not ideal, but at least it is better than it was.

### Documentation

The code is updated with JavaDoc and some inline comments.

### Changelog

✅  Minor bugfix

### Bumping the serialization version id

🟥  Should not be needed, some of the serialized classes are changed, but not the serialization - only added a `normalizer()` method.

t2gran commented:
> * Why don't we change the precision for the paging token instead?

The paging logic is already very complex, and this would add extra complexity. We also want the sorting and itinerary-filer-chain to be deterministic - it needs to use the same normalized values. Rounding as a final step when creating the itineraries is safe, rounding before will lead to rounding errors. The precision OTP aims for is second; Hence we need a higher resolution during computing.

Normailizing the cost and times gives us a small buffer for rounding errors. Of case, if an itinerary has 12000.49 in cost and then, next time the same itinerary get 12000.50, then we would have a problem - but in most cases we would get rid of the precision we do not care about. A better solution is to relax the compare function - but again it will be complex. I think the current solution is the best compromise.

> * Would a cost-class with second precision make more sense to make the precision change from inside raptor to outside raptor more explicit?
> Related to the second point: I'm not sure where the `Cost` class is used, and if there is somewhere a need for it having centi-precision. If there is, it could also be a solution to have two different cost classes, one for each precision.

In Raptor costs are `int` in centi-seconds. We need the higher precision outside Raptor to avoid rounding errors. The Cost class is used as input as well. Also, keep in mind that the `Cost` is not introduced everywhere it should. I am not thrilled about making different types for the same thing, with just different precision. But I guess, a `NormalizedCost` in Leg and Itinerary is a good idea.

tkalvas commented on application/src/main/java/org/opentripplanner/model/plan/Itinerary.java:
```diff
@@ -149,31 +151,38 @@ public ItineraryBuilder copyOf() {
 
   /**
-   * Time that the trip departs.
+   * Time that the trip departs. The time is normalized(rounded to closest second).
```
This is rounded DOWN, not rounded to closest, and if the code is correct, the docstring should match.

But now we have rounding down here, and rounding to closest in the normalize() methods... should the itinerary startTime and endTime also be rounded to closest? I'm not sure if it would break anything either way, but it feels like rounding to closest would be the expected behaviour.

tkalvas commented on application/src/main/java/org/opentripplanner/model/plan/Itinerary.java:
```diff
@@ -149,31 +151,38 @@ public ItineraryBuilder copyOf() {
 
   /**
-   * Time that the trip arrives.
+   * Time that the trip arrives. The value is normalized (rounded to seconds) is requiered for the
```
Here also say rounded down for this behaviour, or then change the code to round to closest and say rounded to closest in docstring.

t2gran commented:
- I have added a `NormalizedCost` type used in Itinerary, Leg uses `int` (bigger refactoring to fix). 
- I fixed the stat/end time for Leg and Itinerary. I normalize these by rounding to closest second.

jessicaKoehnke commented:
I haven't looked at everything yet, but shouldn't `Itinerary#generalizedCostIncludingPenalty` also return a `NormalizedCost`?

jessicaKoehnke commented on application/src/main/java/org/opentripplanner/model/plan/paging/cursor/PageCursorSerializer.java:
```diff
@@ -53,6 +53,9 @@ public static String encode(PageCursor cursor) {
       .withEnum(SORT_ORDER_FIELD, cursor.originalSortOrder());
 
     if (cursor.containsItineraryPageCut()) {
+      // Note! For this to work properly we needt to preserve the same presission for time
```
there are a couple typos: precision, itinerary, normalized

jessicaKoehnke reviewed (commented):
I have just a couple small comments, mostly typos
